### PR TITLE
TransferLearning vs Fine-tuning

### PR DESCRIPTION
**Pull request by Anindyadeep**
# Review checklist

> Don't worry about satisfying all items, it's fine to open a (draft) PR.

- [ ] chapter content
  + [ ] main content focus: recent developments in open source AI
    + general context/background (brief)
    + current pros/cons
    + in-depth insights (not yet widely known)
- [ ] appropriate citations
  + [ ] BibTeX references
  + [ ] Glossary terms
  + [ ] cross-references (figures/chapters)
- [x] images & data not committed to this repo (e.g. use https://github.com/premAI-io/static.premai.io instead)

### CODE REVIEW

**Comment by Anindyadeep** [comment]
Hey, @casperdcl, I am seeing that the link checks is failing for some reason, is this ready for merge?

**Comment by Anindyadeep** [comment]
Hey @casperdcl, so now all the checks are passing. Let me know if this can be merged.

File: fine-tuning.md [hunk 1]
```diff
@@ -19,19 +19,63 @@ Some ideas:
...
+Train on domain-specific data until unfrozen layers converge | yes | yes
+```
+
+### Transfer Learning
+
+The key is that "frozen" layers remain unchanged -- retaining the original abilities of the pre-trained model -- and act as general & robust feature extractors.
+
+```{figure-md} transfer-learning-architecture
+:class: caption
+![](https://static.premai.io/book/transfer_learning.png)
```
Comment (casperdcl) [inline fine-tuning.md:41]: change (consistent capitalisation and hyphens):

- "domain specific input data" -> "domain-specific input data"
- "pretrained model" -> "pre-trained model"
- "Domain specific Model" -> "domain-specific model"

consistency with other diagram:

- pre-trained model: red colour
- domain-specific model: green colour
- add a "prediction" output box

File: fine-tuning.md [hunk 2]
```diff
@@ -19,19 +19,63 @@ Some ideas:
...
+- Limited data: when domain-specific dataset size is small, a large model cannot be trained end-to-end without overfitting. However if the model is mostly a frozen general feature extractor, then the subsequent trainable layers are less likely to overfit.
+- Limited compute and time: retraining a large model from scratch requires a lot of compute resources and time. This is unnecessary if similar performance can be achieved through transfer learning (training just part of a large model).
+
+### Fine-tuning
+
+The key difference here is none (or few) of the pre-trained model's weights are frozen. The pre-training process can be considered an intelligent weight initialisation prior to training on a domain-specific dataset. Essentially, the pre-training will leave the model weights close to a global (general) optimum, while the domain-specific training will find a local (task-specific) optimum.
+
+```{figure-md} fine-tuning-architecture
+:class: caption
+![](https://static.premai.io/book/fine-tuning.png)
```
Comment (casperdcl) [inline fine-tuning.md:62]: change (consistent capitalisation and hyphens):

- "domain specific input data" -> "domain-specific input data"
- "pretrained model" -> "pre-trained model"
- "Frozen" -> "frozen"

consistency with other diagram:

- frozen: red colour
- trainable: green colour

File: fine-tuning.md [hunk 3]
```diff
@@ -19,19 +19,63 @@ Some ideas:
...
+:class: caption
+![](https://static.premai.io/book/transfer_learning.png)
+
+Transfer Learning
+```
+
+**Examples**:
+
+- Computer vision: take [ResNet-50](https://huggingface.co/microsoft/resnet-50) -- which was pretrained on the [ImageNet](TODO:link) dataset -- and replace the last layer with an object-detecting model (e.g. TODO model). This object-detecting model can now be trained to e.g. classify [cats-vs-dogs](https://huggingface.co/datasets/cats_vs_dogs).
+- Natural language processing: take [BERT](https://huggingface.co/google/bert_uncased_L-2_H-768_A-12) -- trained on TODO -- and replace the last layer with TODO. The final layers can be trained on the [tweet sentiment classification dataset](https://huggingface.co/datasets/carblacac/twitter-sentiment-analysis).
```
Comment (casperdcl) [inline fine-tuning.md:49]: added some TODO stubs here

File: index.md [hunk 4]
```diff
@@ -206,6 +206,9 @@ Fair Dealing
 Fair Use
   A doctrine in US law permitting use of {term}`IP` without prior permission (regardless of licence/copyright status) depending on 1) purpose of use, 2) nature of the IP, 3) amount of use, and 4) effect on value {cite}`wiki-fair-use`. See also {term}`fair dealing`.
 
+Fine-tuning
+  TODO: [Fine-tuning](<https://en.wikipedia.org/wiki/Fine-tuning_(machine_learning)>). See also [](fine-tuning) and {term}`transfer learning`.
```
Comment (casperdcl) [inline index.md:210]: concise definition required here

File: index.md [hunk 5]
```diff
@@ -266,6 +269,9 @@ Quantisation
 Token
   A [token](https://learn.microsoft.com/en-us/semantic-kernel/prompt-engineering/tokens) is a "unit of text" for an {term}`LLM` to process/generate. A single token could represent a few characters or words, depending on the tokenisation method chosen. Tokens are usually {term}`embedded <embedding>`.
 
+Transfer Learning
+  TODO. See also {term}`fine-tuning` and [](fine-tuning).
```
Comment (casperdcl) [inline index.md]: concise definition required here.

File: index.md [hunk 6]
```diff
@@ -266,6 +269,9 @@ Quantisation
 Token
   A [token](https://learn.microsoft.com/en-us/semantic-kernel/prompt-engineering/tokens) is a "unit of text" for an {term}`LLM` to process/generate. A single token could represent a few characters or words, depending on the tokenisation method chosen. Tokens are usually {term}`embedded <embedding>`.
 
+Transfer Learning
+  [TODO](https://en.wikipedia.org/wiki/Transfer_learning). See also {term}`fine-tuning` and [](fine-tuning).
```
Comment (casperdcl) [inline index.md:273]: concise definition required here.

File: index.md (hunk 4 above)
Comment (Anindyadeep) [reply index.md:210]: > concise definition required here

So, Should I paste the same definition, that I put inside Fine-tuning section?

```
From [Wikipedia’s](https://en.wikipedia.org/wiki/Fine-tuning_(deep_learning)) definition, Fine-tuning is an approach to transfer learning in which weights of a pre-trained model is trained on a new data. In some case we retrain the whole model on our domain-specific dataset or in other cases, we just fine-tune on only a subset of the layers. Through fine-tuning, we are adapting our existing pre-trained model on a task-specific dataset.
```

**Comment by Anindyadeep** [comment]
Hi @casperdcl, please let me know if this PR is good to get merged :slightly_smiling_face:

File: fine-tuning.md [hunk 7]
```diff
@@ -19,19 +19,79 @@ Some ideas:
...
+Transfer Learning
+```
+
+**Examples**:
+
+- Computer vision: take the [ResNet-50](https://huggingface.co/microsoft/resnet-50) pre-trained on the [ImageNet](https://www.image-net.org/index.php) dataset and replace its last layer with the head of an object-detecting model (such as [Faster R-CNN](https://arxiv.org/abs/1506.01497)). This modified model can now be trained to draw bounding boxes and classify images from the [cats-vs-dogs](https://huggingface.co/datasets/cats_vs_dogs) dataset.
+
+- Natural language processing: take a [BERT](https://huggingface.co/google/bert_uncased_L-2_H-768_A-12) model, that was pre-trained on extensive text data, such as the [BookCorpus dataset](https://huggingface.co/datasets/bookcorpus). Replace BERT's final layer with a simple classifier or Multi-Layer Perceptron (MLP) layers. These final layers can then be trained on the [tweet sentiment classification dataset](https://huggingface.co/datasets/carblacac/twitter-sentiment-analysis) to classify twitter sentiments.
+
+**Use cases**:
```
Comment (nsosio) [inline fine-tuning.md:55]: @Anindyadeep are these use cases or notes? Check also rendering

File: fine-tuning.md [hunk 8]
```diff
@@ -19,19 +19,79 @@ Some ideas:
...
+### Fine-Tuning
+
+From [Wikipedia’s](https://en.wikipedia.org/wiki/Fine-tuning_(deep_learning)) definition, Fine-tuning is an approach to transfer learning in which weights of a pre-trained model is trained on a new data.  In some case we retrain the whole model on our domain-specific dataset or in other cases, we just fine-tune on only a subset of the layers. Through fine-tuning, we are adapting our existing pre-trained model on a task-specific dataset.
+
+```{figure-md} fine-tuning-architecture
+:class: caption
+![](https://static.premai.io/book/fine-tuning.png)
+
+Fine Tuning
+```
```
Comment (nsosio) [inline fine-tuning.md:81]: @Anindyadeep why there are two sections with the same name? In general keep the same structure for both transfer learning and fine-tuning subsections

**Comment by filopedraz** [comment]
@Anindyadeep please handle the last comments from Nico and then we can merge. :)

File: fine-tuning.md (hunk 7 above)
Comment (Anindyadeep) [reply fine-tuning.md:55]: Yeah so the rendering does not work when viewing on readme, rather it works when we build the readmes

File: fine-tuning.md (hunk 8 above)
Comment (Anindyadeep) [reply fine-tuning.md]: Yes actually both had same structure, but there was some redundancy which I removed, hopefully it looks all okay now.

**Review by filopedraz — APPROVED** [review]

**Pull request merged**